File: program/program/implement-binary-search/ImplementBinarySearch.rb
def binary_search(arr, x)
  lower_bound = 0
  upper_bound = arr.length - 1

  while lower_bound <= upper_bound
    mid_point = lower_bound + (upper_bound - lower_bound) / 2

    if arr[mid_point] < x
      lower_bound = mid_point + 1
    elsif arr[mid_point] > x
      upper_bound = mid_point - 1
    else
      return mid_point # x found at mid_point
    end
  end

  return -1 # x does not exist
end

# Example usage
arr = [2, 3, 4, 10, 40]
x = 10

result = binary_search(arr, x)
if result != -1
  puts "Element #{x} is present at index #{result}"
else
  puts "Element #{x} is not present in array"
end

File: program/program/implement-binary-search/implement_binary_search.hs
-- Binary search function
binarySearch :: (Ord a) => [a] -> a -> Int
binarySearch arr x = binarySearchHelper arr x 0 (length arr - 1)

-- Helper function to perform the binary search
binarySearchHelper :: (Ord a) => [a] -> a -> Int -> Int -> Int
binarySearchHelper arr x lowerBound upperBound
  | upperBound < lowerBound = -1  -- x does not exist
  | arr !! midPoint < x      = binarySearchHelper arr x (midPoint + 1) upperBound
  | arr !! midPoint > x      = binarySearchHelper arr x lowerBound (midPoint - 1)
  | otherwise                = midPoint  -- x found at midPoint
  where
    midPoint = lowerBound + (upperBound - lowerBound) `div` 2

-- Example usage
main :: IO ()
main = do
    let arr = [2, 3, 4, 10, 40]
    let x = 10
    let result = binarySearch arr x
    if result /= -1
        then putStrLn $ "Element " ++ show x ++ " is present at index " ++ show result
        else putStrLn $ "Element " ++ show x ++ " is not present in array"
